0867/Original-Submission/Src/quadlog-gampsi-slim/man/glqrc.man
.TH GLQRC 3 "04 November 2003" "Version 1.00"
.\" WARNING: This file was produced automatically from file laguerre/glqrc.f
.\" by fortran-to-man-page.awk on Tue Nov  4 08:23:52 MST 2003.
.\" Any manual changes will be lost if this file is regenerated!
.SH NAME
glqrc \- Gauss-Laguerre logarithmic Quadrature Recursion Coefficients
.\"=====================================================================
.SH SYNOPSIS
Fortran (77, 90, 95, HPF):
.RS
.B f77
.I "[ flags ] file(s) .\|.\|. -L/usr/local/lib -lgjl"
.RS
.nf
.B "SUBROUTINE glqrc (a, b, s, t, alpha, nquad, ierr)"
.B "DOUBLE PRECISION    a(0:MAXPTS), alpha,       b(0:MAXPTS)"
.B "DOUBLE PRECISION    s(0:MAXPTS), t(0:MAXPTS)"
.B "INTEGER             ierr,        nquad"
.fi
.RE
.RE
C (K&R, 89, 99), C++ (98):
.RS
.B cc
.I "[ flags ] -I/usr/local/include file(s) .\|.\|. -L/usr/local/lib -lgjl"
.br
Use
.RS
.B "#include <gjl.h>"
.RE
to get this prototype:
.RS
.nf
.B "void glqrc(fortran_double_precision a_[],"
.B "           fortran_double_precision b_[],"
.B "           fortran_double_precision s_[],"
.B "           fortran_double_precision t_[],"
.B "           const fortran_double_precision * alpha_,"
.B "           const fortran_integer * nquad_,"
.B "           fortran_integer * ierr_);"
.fi
.RE
.RE
.PP
NB: The definition of C/C++ data types
.B fortran_
.IR xxx ,
and the mapping of Fortran external names to C/C++ external names,
is handled by the C/C++ header file.  That way, the same function
or subroutine name can be used in C, C++, and Fortran code,
independent of compiler conventions for mangling of external
names in these programming languages.
.\"=====================================================================
.SH DESCRIPTION
.PP
Compute the recursion coefficients and zeroth and first moments
of the monic polynomials corresponding to the positive weight
function
.PP
.RS
.nf
w(x,\ealpha) = (x \(mi 1 \(mi ln(x)) * exp(\(mix) * x^\ealpha
.fi
.RE
.PP
with recursion relation (n = 0, 1, 2, .\|.\|.)
.PP
.RS
.nf
P_{n+1}^\ealpha(x) = (x \(mi B_n^\ealpha) * P_n^\ealpha(x)
            \(mi A_n^\ealpha * P_{n\(mi1}^\ealpha(x)
.fi
.RE
.PP
and initial conditions
.PP
.RS
.nf
P_{\(mi1}^\ealpha(x) = 0
P_{0}^\ealpha(x) = 1
.fi
.RE
.PP
Except in the weight function, the superscripts indicate
dependence on \ealpha, NOT exponentiation.
.PP
The required moments are:
.PP
.RS
.nf
T_n^\ealpha = \eint_0^\einfty w(x,\ealpha) (P_n^\ealpha(x))^2 dx
S_n^\ealpha = \eint_0^\einfty w(x,\ealpha) (P_n^\ealpha(x))^2 x dx
.fi
.RE
.PP
From these moments, the recursion coefficents are computed as:
.PP
.RS
.nf
A_n^\ealpha = T_n^\ealpha / T_{n\(mi1}^\ealpha
B_n^\ealpha = S_n^\ealpha / T_n^\ealpha
.fi
.RE
.PP
On entry:
.PP
.RS
.TP \w'\fBalpha\fP\&'u+2n
\fBalpha\fP\&
Power of x in the integrand (\fBalpha\fP\& > \(mi1).
.TP
\fBnquad\fP\&
Number of quadrature points to compute. It
must be less than the limit MAXPTS defined
in the header file,
.I maxpts.inc.
The default value chosen there should be large
enough for any realistic application.
.RE
.PP
On return:
.PP
.RS
.TP \w'\fBa\fP\&(0.\|.\fBnquad\fP\&)'u+2n
\fBa\fP\&(0.\|.\fBnquad\fP\&)
Recursion coefficients: \fBa\fP\&(n) = A_n^\ealpha.
.TP
\fBb\fP\&(0.\|.\fBnquad\fP\&)
Recursion coefficients: \fBb\fP\&(n) = B_n^\ealpha.
.TP
\fBs\fP\&(0.\|.\fBnquad\fP\&)
First moments: \fBs\fP\&(n) = S_n^\ealpha
.TP
\fBt\fP\&(0.\|.\fBnquad\fP\&)
Zeroth moments: \fBt\fP\&(n) = T_n^\ealpha
.TP
\fBierr\fP\&
Error indicator:
.nf
= 0 (success),
1 (eigensolution could not be obtained),
2 (destructive overflow),
3 (\fBnquad\fP\& out of range),
4 (\fBalpha\fP\& out of range).
.fi
.RE
.\"=====================================================================
.SH AUTHORS
The algorithms and code are described in detail in
the paper
.RS
.I "Fast Gaussian Quadrature for Two Classes of Logarithmic Weight Functions"
.RE
in ACM Transactions on Mathematical Software,
Volume ??, Number ??, Pages ????--???? and
????--????, 20xx, by
.RS
.nf
Nelson H. F. Beebe
Center for Scientific Computing
University of Utah
Department of Mathematics, 110 LCB
155 S 1400 E RM 233
Salt Lake City, UT 84112-0090
Tel: +1 801 581 5254
FAX: +1 801 581 4148
Email: \fCbeebe@math.utah.edu\fP, \fCbeebe@acm.org\fP, \fCbeebe@computer.org\fP
WWW URL: \fChttp://www.math.utah.edu/~beebe\fP
.fi
.RE
and
.RS
.nf
James S. Ball
University of Utah
Department of Physics
Salt Lake City, UT 84112-0830
USA
Tel: +1 801 581 8397
FAX: +1 801 581 6256
Email: \fCball@physics.utah.edu\fP
WWW URL: \fChttp://www.physics.utah.edu/people/faculty/ball.html\fP
.fi
.RE
.\"==============================[The End]==============================
.\"=====================================================================
.\" This is for GNU Emacs file-specific customization:
.\" Local Variables:
.\" fill-column: 50
.\" End:


Changes
April 22 2021
   * Added 1013-1014 from Volume 47, number 1, (January 2021)
   * Added 1012 from Volume 46, number 4, (November 2020)
October 19 2020
   * Added 1011 from Volume 46, number 3, (September 2020)
August 7 2020
   * Added 1003-1006 from Volume 46, number 1, (April 2020)
   * Added 1007-1010 from Volume 46, number 2, (June 2020)
January 12 2020
   * Updated 999 at the request of the author -- updated user manual supplied.
January 7 2020
   * Added 999-1002 from Volume 45, number 4, (December 2019)
October 13 2019
   * Added 995-998 from Volume 45, number 3, (August 2019)
August 8 2019
   * Added 993 and 994 from Volume 45, number 2, (June 2019)
August 7 2019
   * Added 991 and 992 from Volume 45, number 1, (February 2019)
February 1 2019
   * Added 989 and 990 from Volume 44, number 3, (August 2018)
May 8 2018
   * Added 987 and 988 from Volume 44, number 3, (April 2018)
October 13 2017
   * Added 986 from Volume 44, number 2, (October 2017)
   * Corrected publication month for Algs 979-985 (now Oct).
September 30 2017
   * Added 979-985 from Volume 44, number 2, (September 2017)
   * Added 975-978 from Volume 44, number 1, (July 2017)
April 21 2017
   * Added 973, 974 and Remark on 936 from Volume 43, number 4 
     (March 2017)
February 3 2017
   * Added 968-972 from Volume 43, number 3 (January 2017)
September 16 2016
   * Added 965 from Volume 43, number 1 (August 2016)
   * Added 966, 967 from Volume 43, number 2 (September 2016)
   * Removed files 620 and 620.bib (all other algorithm files
     are only available as compressed files.
   * Replaced index.gz and index.html.gz by .zip files
September 14 2016:
   * Added 100.zip and 101.zip. These files are identical as
   the two algorithms are inter-related and the driver program
   uses both codes.
August 30 2016:
   * Added 963-964 from Volume 42, number 4 (July 2016)
July 23 2016
   * Updated contact information on CALGO web pages.
June 19 2016:
   * Added 960-962 from Volume 42, number 3 (June 2016)
   * Updated 916 to include the code submitted with Remark 
     in Volume 42, number 3 (Jun 2016)
June 15 2016:
   * Added 957 from Volume 42, number 1 (February 2016)
   * Added 958-959 from Volume 42, number 2 (June 2016)
May 30 2016
   * Added 955-956 from Volume 42, number 1 (February 2016)
May 27 2016:
   * Added new code supplied as Remark to Algorithm 673
     appearing in Volume 42, number 1.
December 22 2015:
   * Added 947-954 from Volume 41, numbers (1-4).
August 29 2015:
   * Mended a number of odd/broken compressed files that were on
     the CALGO site:
       755.gz, 756.gz, 823.gz replaced by zip files
       709.gz, 709a.gz, 709b.gz replaced by 709.zip
       573.gz, 573d.gz, 573s.gz replaced by 573.zip
       358.gz, 805.gz, 830.gz, 832.zip all updated
       620.zip added -- this contains all the 620 components.
September 26 2014
   * Updated 897 following accepted remark.
September 1 2014
   * Added 943-946 from Volume 40, number 4 (June 2014)
May 17 2014
   * Fixed problem with Algorithm 917 missing from indexes.
May 5 2014
   * Added 935-938 from Volume 40, number 2 (February 2014)
   * Added 939-942 from Volume 40, number 3 (April 2014)
October 9 2013
   * Added 929, 930 from volume 39, number 4 (July 2013)
   * Added 931-934 from volume 40, number 1 (October 2013)
July 18 2013
   * Added 928 from volume 39, number 3 (May 2013)
   * Updated 905 from code supplied by author thackerw@winthrop.edu
March 3 2013
   * Added 926-927 from volume 39, number 2 (February 2013)
January 25 2013
   * Replaced 923 after editing Licence file -- author request
January 3 2013
   * Added 924-925 from volume 39, number 1 (November 2012)
September 10 2012
   * Added 920-923 from volume 38, number 4 (August 2012)
June 17 2012
   * Added 917-919 from volume 38, number 3 (April 2012)
March 14 2012
   * Corrected typo in index files (Alg 916)
February 29 2012
   * Replaced 916 -- old version of code uploaded in error.
February  5 2012
   * Added 916 from Volume 38, number 2 (December 2012)
December 30 2011
   * Replaced 905 due to corrupt blas.f file in the original (pointed
     out by Ronald.M.Bloom@aero.org).
December 7 2011
   * Added 913-915 from Volume 38, number 1 (November 2011)
   * Updates 778 as per Remark in Volume 38, No. 1
April 12 2011
   * Added 909-911 from volume 37, number 4 (April 2011)
October 24 2011
   * Changed Algorithm 419 in line with changes from Remark on Algorithm 419 by
     David H. Withers CACM (March 1974) Vol 17 No 3 p. 157
     Pointed out by j.a.horsmeier@wanadoo.nl (Jos)
October 11 2010
   * Added 904-908 from volume 37, number 3 (September 2010)
May  20 2010
   * Added 902, 903 from volume 37, number 2 (April 2010)
Feb  27 2010
   * Added 898, 899, 900, 901 from volume 37, number 1 (January 2010)
July 29 2009
   * Added 895, 896, 897 from volume 36, number 3 (July 2009)
Apr  6 2009
   * Added 894 from volume 36, number 2 (March 2009)
Mar  30 2009
   * Added 891, 892, 893 from volume 36, number 1 (March 2009)
Feb  12 2009
   * Added 890 from volume 35, number 4 (February 2009)
Jan   6 2009
   * Added 885-889 from volume 35, number 3 (October 2008)
Sept  4 2008
   * Added 881-884 from volume 35, number 2 (July 2008)
July 25 2008
   * Added 876-877 from volume 34, number 4 (July 2008)
   * Added 878-880 from volume 35, number 1 (July 2008)
June 11 2008
   * Added 873-875 from volume 34, number 2 and 3 (March and May 2008)
Feb   8 2008
   * Added 867-869 from volume 33, numbers 3 and 4 (August 2007)
   * Added 870-872 from volume 34, number 1 (January 2008)
Aug   3 2007
   * Added 863-866 from volume 33, numbers 1 and 2.
July  3 2007
   * Updated 620 to include GAMS classifications to algorithms 814-862 inc.
Jan  10 2007
   * Updated 644 as a result of an accepted remark by Kodama
Jan   5 2007
   * Added 857-862 from volume 32, 4 (December 2006)
Nov  21 2006
   * Added 358 -- donated by John Burkardt <burkardt@csit.fsu.edu>.
Oct   5 2006
   * Added 854 from volume 32, 2 (June 2006)

0867/Original-Submission/Src/quadlog-gampsi-slim/laguerre/glqrc.f
      SUBROUTINE  glqrc (a, b, s, t, alpha, nquad, ierr)
************************************************************************
*     (Gauss-Laguerre Logarithmic Quadrature Recursion Coefficients)
*
*     Compute the recursion coefficients and zeroth and first moments
*     of the monic polynomials corresponding to the positive weight
*     function
*
*         w(x,\alpha) = (x - 1 - ln(x)) * exp(-x) * x^\alpha
*
*     with recursion relation (n = 0, 1, 2, ...)
*
*         P_{n+1}^\alpha(x) = (x - B_n^\alpha) * P_n^\alpha(x)
*                             - A_n^\alpha * P_{n-1}^\alpha(x)
*
*     and initial conditions
*
*         P_{-1}^\alpha(x) = 0
*         P_{0}^\alpha(x) = 1
*
*     Except in the weight function, the superscripts indicate
*     dependence on \alpha, NOT exponentiation.
*
*     The required moments are:
*
*         T_n^\alpha = \int_0^\infty w(x,\alpha) (P_n^\alpha(x))^2 dx
*         S_n^\alpha = \int_0^\infty w(x,\alpha) (P_n^\alpha(x))^2 x dx
*
*     From these moments, the recursion coefficents are computed as:
*
*         A_n^\alpha = T_n^\alpha / T_{n-1}^\alpha
*         B_n^\alpha = S_n^\alpha / T_n^\alpha
*
*     On entry:
*
*          alpha           Power of x in the integrand (alpha > -1).
*
*          nquad           Number of quadrature points to compute.  It
*                          must be less than the limit MAXPTS defined
*                          in the header file, maxpts.inc.  The default
*                          value chosen there should be large enough
*                          for any realistic application.
*
*     On return:
*
*          a(0..nquad)     Recursion coefficients: a(n) = A_n^\alpha.
*
*          b(0..nquad)     Recursion coefficients: b(n) = B_n^\alpha.
*
*          s(0..nquad)     First moments: s(n) = S_n^\alpha
*
*          t(0..nquad)     Zeroth moments: t(n) = T_n^\alpha
*
*          ierr            Error indicator:
*                          = 0  (success),
*                            1  (eigensolution could not be obtained),
*                            2  (destructive overflow),
*                            3  (nquad out of range),
*                            4  (alpha out of range).
*
*     [27-Oct-2003] -- new code in separate subroutine to allow access
*                      to recursion coefficients and moments
*     [18-Mar-2000] -- original code embedded in glqfd()
************************************************************************
*
*     External functions
*
      EXTERNAL            dgamma,      dpsi,        dvsum
*
      DOUBLE PRECISION    dgamma,      dpsi,        dvsum
*
*     Parameter variables
*
      DOUBLE PRECISION    ONE
      PARAMETER           (ONE = 1.0d+00)
      DOUBLE PRECISION    TWO
      PARAMETER           (TWO = 2.0d+00)
      DOUBLE PRECISION    ZERO
      PARAMETER           (ZERO = 0.0d+00)
*
      INCLUDE 'ecodes.inc'
      INCLUDE 'maxpts.inc'
*
*     Argument variables
*
      DOUBLE PRECISION    a(0:MAXPTS), alpha,       b(0:MAXPTS)
      DOUBLE PRECISION    s(0:MAXPTS), t(0:MAXPTS)
*
      INTEGER             ierr,        nquad
*
*     Local variables
*
      DOUBLE PRECISION    deltaw(MAXPTS),           deltax(MAXPTS)
      DOUBLE PRECISION    p(MAXPTS),   pm(MAXPTS),  pp(MAXPTS)
      DOUBLE PRECISION    ppm(MAXPTS), sn,          tmp,         tmpp
      DOUBLE PRECISION    tn,          w(MAXPTS),   wxm1dw(MAXPTS)
      DOUBLE PRECISION    x(MAXPTS)
*
      INTEGER             m,           n,           nqp2
*
*     Accurate computation of S(nquad) requires a quadrature order at
*     least TWO higher than requested, since S(n) has an extra x
*     factor, and the weight function has an x factor, giving an
*     integrand of polynomial order 2*n + 2 = 2*(n + 1).
*
      nqp2 = nquad + 2
      CALL glqfd(x, w, deltaw, deltax, alpha, nqp2, ierr)
      IF (ierr .NE. eokay) RETURN
*
*     Compute the quadrature weights needed in glqli() and for the
*     initial moments t(0) and s(0).
*
      DO 100 n = 1, nqp2
          wxm1dw(n) = (x(n) - ONE)*w(n) - deltaw(n)
*         g(n) = wxm1dw(n) * x(n)
  100 CONTINUE
*
*     We could compute the initial moments from (exact to
*     machine-precision) quadrature, but it is more accurate to use
*     the analytic expressions in dgamma() and dpsi() given below.
*
*     t(0) = \int_0^\infty w(x,\alpha) (P_0^\alpha(x))^2 dx
*          = \int_0^\infty w(x,\alpha) dx
*          = sum of weights wxm1dw(*)
*          = gamma(\alpha + 1) * (\alpha - psi(\alpha + 1))
*
*     s(0) = \int_0^\infty w(x,\alpha) (P_0^\alpha(x))^2 x dx
*          = \int_0^\infty w(x,\alpha) x dx
*          = sum(i=1:n)(g(i) - deltax(i))
*          = gamma(\alpha + 2) * (\alpha + 1 - psi(\alpha + 2))
*
*     t(0) = dvsum(wxm1dw,nqp2)
*     g(0) = -dvsum(deltax, nqp2)
*     s(0) = dvsum(g, 1 + nqp2)
      t(0) = dgamma(alpha + ONE) * (alpha - dpsi(alpha + ONE))
      s(0) = dgamma(alpha + TWO) * (alpha + ONE - dpsi(alpha + TWO))
*
      a(0) = t(0)
      b(0) = s(0) / t(0)
*
*     Initialize monic polynomial starting values.
*
      DO 200 n = 1, nqp2
          p(n) = ONE
          pm(n) = ZERO
          ppm(n) = ZERO
          pp(n) = ZERO
  200 CONTINUE
*
*     Because a(0), b(0), s(0), and t(0) have already been determined
*     by accurate analytic formulas, on the n = 0 iteration, we avoid
*     replacing them by slightly less accurate values determined by
*     numerical quadrature.
*
      DO 400 n = 0, nquad
          CALL glqli(p, pp, deltax, wxm1dw, x, nqp2, sn, tn)
          IF (n .GT. 0) THEN
              s(n) = sn
              t(n) = tn
              a(n) = t(n) / t(n - 1)
              b(n) = s(n) / t(n)
          END IF
          DO 300 m = 1, nqp2
              tmp = p(m)
              tmpp = pp(m)
              p(m) = (x(m) - b(n))*tmp - a(n)*pm(m)
              pp(m) = ((x(m) - b(n))*tmpp + tmp) - a(n)*ppm(m)
              pm(m) = tmp
              ppm(m) = tmpp
  300     CONTINUE
  400 CONTINUE
*
      END
